lib/logstash/filters/preprocessor.rb

```diff
@@ -0,0 +1,201 @@
+    if match.nil?
+      nil
+    else
+      # Gets content
+      content = match[@content_specifier]
+
+      # Mask content and return
+      processed_line = mask_log_event(content)
+      processed_line.strip.split
+    end
+  end
+
+  # Processes an array of tokens to upload single grams, digrams, and trigrams to the @gram_dict.
+  #
+  # This method iterates through each token in the array. For each token, it uploads the token as a single gram.
+  # Additionally, if the current token is not the first in the array, it creates and uploads a digram using the current
+  # and previous token.
+  # If the token is at least the third in the array, the method also creates and uploads a trigram using the current
+  # token and the two preceding it.
+  # The tokens in digrams and trigrams are separated by a defined separator (`token_seperator`).
+  #
+  # @param tokens [Array<String>] an array of string tokens to be processed
+  # @return [void] this method does not return a value but updates the @gram_dict object.
+  def upload_grams_to_gram_dict(tokens)
+    token_seperator = '^'
+
+    # Iterate across all tokens
+    tokens.each_with_index do |token, index|
+      # Upload single gram
+      @gram_dict.single_gram_upload(token)
+
+      # If possible, upload a digram
+      if index.positive?
+        first_token = tokens[index - 1]
+        digram = first_token + token_seperator + token
+        @gram_dict.double_gram_upload(digram)
+      end
+
+      # If possible, upload a trigram
+      next unless index > 1
+
+      first_token = tokens[index - 2]
+      second_token = tokens[index - 1]
+      trigram = first_token + token_seperator + second_token + token_seperator + token
+      @gram_dict.tri_gram_upload(trigram)
+    end
+  end
+
+  # Processes a given log event by tokenizing it and updating the gram dictionary.
+  #
+  # This method first calls the `token_splitter` method to split the log event into tokens based on the
+  # pre-configured format.
+  # The tokens are then passed to the `upload_grams_to_gram_dict` method, which iteratively uploads single grams,
+  # digrams, and trigrams to the `@gram_dict`.
+  #
+  # The process involves two primary steps: tokenization and dictionary updating.
+  # Tokenization is done based on the log format and involves masking sensitive information before splitting.
+  # Each token, digram, and trigram found in the log event is then uploaded to the gram dictionary, enhancing the
+  # dictionary's ability to process future log events.
+  #
+  # @param log_event [String] the log event to be processed
+  # @return [void] this method does not return a value but updates the @gram_dict object by adding new entries
+  #  based on the tokens extracted from the log event.
+  def process_log_event(log_event)
+    # Split log event into tokens
+    tokens = token_splitter(log_event)
+
+    # Upload tokens to gram dictionary if exists
+    if tokens != nil
+      upload_grams_to_gram_dict(tokens)
```

Comment:
future PR: if gramdict just lives in here, maybe we can refactor out preprocessing into a separate class and just setup all log parsing in here (with a single parser class and single gramdict class that can reference each other) - but we can discuss in more detail at our next meeting